File: euclidean_primer/euclidean_primer__rsa_verification.py
# 実行時間 10ms
def calcN(n):
    i = 2
    l = []

    while i * i <= n:
        while n % i == 0:
            l.append(i)
            n //= i
        i += 1
    if n != 1:
        l.append(n)
    return l

def extGcd(a, b):
    if b != 0:
        c, y, x = extGcd(b, a % b)

        y -= a // b * x
        return c, x, y
    return a, 1, 0

n, e, E = [3995747143, 3007, 602607029]
pr = calcN(n)
nd = (pr[0] - 1) * (pr[1] - 1)
b2 = f"{bin(pow(E, (nd + extGcd(e, nd)[1]) % nd, n))[2:]}"

while len(b2) % 7 != 0:
    b2 = "0" + b2
for i in range(0, len(b2), 7):
    b10 = int(b2[i : i + 7], 2)

    if b10 != 0:
        print(chr(b10), end="")
print()

File: euclidean_primer/euclidean_primer__rsa_verification.cs
// 実行時間 20ms
using System;
using static System.Console;

internal class Program
{
    static void Main()
    {
        long[] neE = { 3995747143, 3007, 602607029 };
        long[] pr = CalcN( neE[0] );
        long nd = ( pr[0] - 1 ) * ( pr[1] - 1 );
        long x = 1;
        long y = 1;

        WriteLine( string.Join( " ", neE ) );
        ExtGcd( neE[1], nd, ref x, ref y );

        string b2 = Convert.ToString( ModPow( neE[2], ( nd + x ) % nd, neE[0] ), 2 );

        while( b2.Length % 7 != 0 ) {
            b2 = "0" + b2;
        }
        for( int i = 0; i < b2.Length; i += 7 ) {
            int b10 = Convert.ToInt32( b2.Substring( i, 7 ), 2 );

            if( b10 != 0 ) {
                Write( Convert.ToChar( b10 ) );
            }
        }
        WriteLine();
        return;
    }

    static long[] CalcN( long n )
    {
        var l = new System.Collections.Generic.List<long>();

        for( long i = 2; i * i <= n; i++ ) {
            while( n % i == 0 ) {
                l.Add( i );
                n /= i;
            }
        }
        if( n != 1 ) {
            l.Add( n );
        }
        return l.ToArray();
    }

    static long ExtGcd( long a, long b, ref long x, ref long y )
    {
        long c = a;

        if( b != 0 ) {
            c = ExtGcd( b, a % b, ref y, ref x );
            y -= a / b * x;
        } else {
            x = 1;
            y = 0;
        }
        return c;
    }

    static long ModPow( long a, long b, long m )
    {
        ulong la = (ulong)a;
        ulong lb = (ulong)b;
        ulong lm = (ulong)m;
        ulong c = 1;

        while( 0 < lb ) {
            if( ( lb & 1 ) == 1 ) {
                c = c * la % lm;
            }
            la = la * la % lm;
            lb >>= 1;
        }
        return (long)c;
    }
}
